### DISCUSSION TITLE
Support function input in useHydrateAtoms

### DISCUSSION THREAD

**Discussion by yf-yang** [pmndrs/jotai#2215] [Ideas] [answered]
Just start an independent discussion thread and copy the ideas from https://github.com/jotaijs/jotai-scope/issues/12#issuecomment-1780412687

Thoughts:

The equivalent of

``` js
const initializer = () => { return someHeavyWorks() };
useState(initializer);
```

for globally unique atom, it should be:

``` js
const initializer = () => { return someHeavyWorks() };
const anAtom = atom(initializer());
// not: const anAtom = atom(initializer);
```

For initialize atom in component, it should be

``` js
const initializer = () => { return someHeavyWorks() };
useHydrateAtoms([[anAtom, initializer]]);
```

Therefore, can we allow function parameter in useHydrateAtoms?

**Comment by dai-shi** [comment]
I think I get what generally you want, but I don't think I get the analogy:
```js
const anAtom = atom(initializer());
// 👆 seems to lead to 👇 
useHydrateAtoms([[anAtom, initializer()]]);
```

**Reply by yf-yang** [reply]
Yep, but here initializer get called during every render, and the hydratedSet internally checks whether the atom is hydrated, so I consider supporting that within useHydrateAtoms, so we don't need to call useMemo outside of useHydrateAtoms.

**Reply by dai-shi** [reply, marked as answer]
Oh, I see what you mean. `someHeavyWorks`

Yeah, I'd suggest `useMemo` in such a case, or `useRef`. You could create a wrapper around useHydrateAtoms. `useLazyHydrateAtoms`.

Apart from useHydrateAtoms, the fact is lazy evaluation of atoms isn't well designed. It's partly due to my oversight. I would design/implement it slightly differently, but I don't think we can do that now as it's breaking.
